### PR TITLE
feat: rewind to user message

### PR DESCRIPTION
**Pull request by lukealvoeiro**
- Removes retry logic from Exchange.generate, as this may result in undefined behavior
- Adds a rewind_to_last_user_message method on the Exchange

### CODE REVIEW

File: src/exchange/exchange.py [hunk 1]
```diff
@@ -77,38 +80,17 @@ def add(self, message: Message) -> None:
...
+            message, usage = self.provider.complete(
+                self.model,
+                self.system,
+                messages=self.messages,
+                tools=self.tools,
+            )
+        except HTTPStatusError as e:
+            # in the future, if models are hosted on machines we may need to add
+            # other types of exceptions here
+            raise FailedToGenerateMessageError(f"Failed to generate message: {e}") from e
```
Comment (baxen) [inline src/exchange/exchange.py:93]: I'd suggest leaving this as the original HTTP Status Error? that has more actionable information when caught downstream or for debugging (like the request etc)

File: src/exchange/exchange.py [hunk 2]
```diff
@@ -337,6 +319,19 @@ def prepend_checkpointed_message(self, message: Message, token_count: int) -> No
         )
         self.checkpoint_data.message_index_offset = new_index
 
+    def rewind_to_last_user_message(self) -> None:
```
Comment (baxen) [inline src/exchange/exchange.py:322]: nit: rewind_to sounds like the last user message is included, maybe `rewind_past_`

File: tests/test_exchange.py [hunk 3]
```diff
@@ -689,67 +689,28 @@ def test_generate_successful_response_on_first_try(normal_exchange):
...
-    with pytest.raises(Exception) as e:
-        ex.generate()
-    assert str(e.value) == "Failed to generate the next message."
+    ex.generate()
+
+    # testing if it works with a non-user text message at the end
+    ex.rewind_to_last_user_message()
+
+    assert len(ex.messages) == 2
+    assert len(ex.checkpoint_data.checkpoints) == 2
```
Comment (baxen) [inline tests/test_exchange.py:716]: probably worth also adding tool use/results into this test?

**Review by baxen — APPROVED** [review]
Looks great!

File: src/exchange/exchange.py (hunk 1 above)
Comment (lukealvoeiro) [reply src/exchange/exchange.py:93]: Ok, I figured we might want to handle all `generate` errors similarly, but not attached to it.

File: src/exchange/exchange.py (hunk 2 above)
Comment (lukealvoeiro) [reply src/exchange/exchange.py:322]: 👍🏼

File: tests/test_exchange.py (hunk 3 above)
Comment (lukealvoeiro) [reply tests/test_exchange.py:716]: 👍🏼

**Review by andrewblane — APPROVED** [review]

**Pull request merged**